spec/predicates/deployment.md

```diff
@@ -0,0 +1,477 @@
+
+### Verification
+
+#### Configuration
+Verification of a deployment attestation is typically performed by an admission controller prior to deploying an artifact / container.
+The verification configuration MUST be done out-of-band and contain the following pieces of information:
+
+1. Required: The "trusted roots". A trusted root defines an entity that is trusted to generate attestations. A trusted root MUST be configured with the following pieces of information:
+    - Required: The unique identity of the attestation generator. The identity may be a cryptographic public key, an identity in an x509 certificate, etc.
+    - Required: Which scope types the attestation generator is authoritative for.
+2. Optional: Required scopes, which is a set of mandatory scope types that MUST be non-empty for verification to pass. Images MUST have attestation(s) over each scope type in the set in order to be admitted. Required scopes are necessary in an attestation, but not sufficient; other scopes present in the attestation MUST match the current environment in order for it to be considered valid.
+3. Optional: Required URI for each scope type that identifies resources implicitely (see [Schema](#schema)).
+
+#### Logic
+
+Verification happens in two phases:
+
+1. Attestation authenticity verification. It takes as input an artifact, an attestation, an attestation signature and the trusted roots. If verification passes, it outputs the attestation's intoto statement. If verification fails, the attestation is considered invalid and MUST be rejected. Using the trusted roots, this phase verifies:
+    - The attestation signature.
+    - The authoritative scopes present in the attestation. If a scope type is non-empty and the generator is _not_ authoritative for the scope type, verification MUST fail. If a scope type is unrecognized or not supported by the verifier, verification MUST fail.
+    - Required scopes are present in the attestation.
+2. Scope match verification. It takes as input the intoto statement from the previous phase. For scope types that identify a resource explicitly (see [Schema](#schema)), the verifier matches each scope value against its corresponding environment value where the artifact is to be deployed
+(e.g., a service account, a pod ID). For scope types that identify a resource implicitly via an authorization URI (see [Schema](#schema)), the verifier matches the value against the URI in the configuration (See [Configuration](#configuration)). Non-empty fields add constraints to the protection scope and are _always_ interpreted as a logical "AND". The verifier MUST compare each scope value to its expected value using an equality comparison. If the values are all equal, verification passes. Otherwise, it MUST fail. Unset scopes (either a scope type with an empty value or a non-present scope) are interpreted as "any value" and are ignored.
+
+### Supported Scopes
```

Comment:
Note: I want to separate versioning of the attestation and versioning of scopes, so that the attestation version can remain the same while the scopes evolve. I probably need to add a sentence about this; or we create a subfolder with protection scope files for each environment (GCP, k8, Spiffe, SELinux, etc)

Reply:
I guess the trick is that we want to list some scopes here and we don't want to have to rev the predicate type when adding new scopes?

Since users can already add scopes of their own, outside this spec, would it be fine to just increase the minor version when new scopes are added?  Users shouldn't be impacted by new scopes or new versions being added so the major version (and thus the predicateType string) wouldn't have to change.

Alternatively maybe a new subdirectory could be added where the scopes definitions could be kept?

Reply:
> I guess the trick is that we want to list some scopes here and we don't want to have to rev the predicate type when adding new scopes?

correct

> Since users can already add scopes of their own, outside this spec, would it be fine to just increase the minor version when new scopes are added? Users shouldn't be impacted by new scopes or new versions being added so the major version (and thus the predicateType string) wouldn't have to change.

From a verifier perspective, a new scope is a breaking change, I think: if a producer starts adding the scope, a verifier would fail since it would not be authoritative for it. Wdut?

> Alternatively maybe a new subdirectory could be added where the scopes definitions could be kept?

That's currently the best option imo. This way we can version each scope in their individual files (k8, GCP, etc)

Reply:
> From a verifier perspective, a new scope is a breaking change, I think: if a producer starts adding the scope, a verifier will fail. Wdut?

The spec already allows producers to add their own custom scopes if they want. So I figure the producer of the attestations and the downstream verifiers must already be in sync here.

Documenting a scope that verifiers could use isn't the same as adding it to the an attestation when it's produced where you'd better be sure the downstream verifiers knows what to do in the first place.

That's especially true if these wind up getting used outside of Kubernetes.  An Android phone won't have a spiffed ID?  or a gke namespace?

Reply:
> The spec already allows producers to add their own custom scopes if they want. So I figure the producer of the attestations and the downstream verifiers must already be in sync here.

Yes they are in sync. All I'm saying is that, e.g., a GCP scope could have a v2 that adds new fields, and the verifier needs to be able to claim they support scope@v2, not just @v1.

Maybe what you're saying is that someone could create a custom scope that extends an existing scope? In which case we don't care about versioning at all, because we just treat them as custom scope. Is that what you meant? 

It's still useful to let others adopt this new extension, which is what the @v2 would help with. It also lets cloud providers who offer such features to say what (claims / version) they support out of the box.

TL;DR: I don't think I disagree with what you're saying, but I see some value in versioning the scope types separately.

Reply:
Resolving. Please re-open if you think we need more discussion. Thanks!